query/src/main/scala/filodb/query/exec/SetOperatorExec.scala

```diff
@@ -0,0 +1,142 @@
+                                ignoring: Seq[String]) extends NonLeafExecPlan {
+  require(cardinality == Cardinality.ManyToMany, "set operations must only use many-to-many matching")
+
+  require(on == Nil || ignoring == Nil, "Cannot specify both 'on' and 'ignoring' clause")
+  require(!on.contains("__name__"), "On cannot contain metric name")
+
+  val onLabels = on.map(Utf8Str(_)).toSet
+  val ignoringLabels = ignoring.map(Utf8Str(_)).toSet + "__name__".utf8
+  // if onLabels is non-empty, we are doing matching based on on-label, otherwise we are
+  // doing matching based on ignoringLabels even if it is empty
+  val onMatching = onLabels.nonEmpty
+
+  def children: Seq[ExecPlan] = lhs ++ rhs
+
+  protected def schemaOfCompose(dataset: Dataset): ResultSchema = lhs(0).schema(dataset)
+
+  protected def args: String = s"binaryOp=$binaryOp, on=$on, ignoring=$ignoring"
+
+  protected[exec] def compose(dataset: Dataset,
+                              childResponses: Observable[(QueryResponse, Int)],
+                              queryConfig: QueryConfig): Observable[RangeVector] = {
+    val taskOfResults = childResponses.map {
+      case (QueryResult(_, _, result), i) => (result, i)
+      case (QueryError(_, ex), _)         => throw ex
+    }.toListL.map { resp =>
+      require(resp.size == lhs.size + rhs.size, "Did not get sufficient responses for LHS and RHS")
+      val lhsRvs = resp.filter(_._2 < lhs.size).flatMap(_._1)
+      val rhsRvs = resp.filter(_._2 >= lhs.size).flatMap(_._1)
+
+      val results: List[IteratorBackedRangeVector] = binaryOp  match {
+        case LAND => setOpAnd(lhsRvs, rhsRvs)
+        case LOR => setOpOr(lhsRvs, rhsRvs)
+        case LUnless => setOpUnless(lhsRvs, rhsRvs)
+        case _ => throw new IllegalArgumentException("requirement failed: " + "Only and, or and unless are supported ")
+      }
+
+      Observable.fromIterable(results)
+    }
+    Observable.fromTask(taskOfResults).flatten
+  }
+
+  private def joinKeys(rvk: RangeVectorKey): Map[Utf8Str, Utf8Str] = {
+    if (onLabels.nonEmpty) rvk.labelValues.filter(lv => onLabels.contains(lv._1))
+    else rvk.labelValues.filterNot(lv => ignoringLabels.contains(lv._1))
+  }
+
+  private def setOpAnd(lhsRvs: List[SerializableRangeVector]
+                       , rhsRvs: List[SerializableRangeVector]): List[IteratorBackedRangeVector] = {
+    val rhsKeysSet = new mutable.HashSet[Map[Utf8Str, Utf8Str]]()
+    var result = new ListBuffer[IteratorBackedRangeVector]()
+    rhsRvs.foreach { rv =>
+      val jk = joinKeys(rv.key)
+      if (!jk.isEmpty)
+        rhsKeysSet += jk
+    }
+
+    lhsRvs.foreach { lhs =>
+      val jk = joinKeys(lhs.key)
+      // Add range vectors from lhs which are present in lhs and rhs both
+      // Result should also have range vectors for which rhs does not have any keys
+      if (rhsKeysSet.contains(jk) || rhsKeysSet.isEmpty) {
+        result += IteratorBackedRangeVector(lhs.key, lhs.rows)
```

Comment:
No need to create new RV. Can simply used `lhs`. Same comment on a few other places.

Reply:
I missed that. Made the change